libs/application-generic/src/instrumentation/instrumentation.decorator.ts

```diff
@@ -0,0 +1,90 @@
+import 'reflect-metadata';
+
+function copyMetadata(source: any, target: any): void {
+  const result = Reflect.getMetadataKeys(source);
+
+  for (const key of result) {
+    Reflect.defineMetadata(key, Reflect.getMetadata(key, source), target);
+  }
+}
+
+type InstrumentationOptions = {
+  transactionName?: string;
+  buildTransactionIdSuffix?: (...args: any[]) => string;
+};
+
+export const InstrumentUsecase = ({
+  transactionName = '',
+  buildTransactionIdSuffix,
+}: InstrumentationOptions = {}): any =>
+  instrumentationWrapper({
+    transactionName,
+    instrumentationType: 'Usecase',
+    buildTransactionIdSuffix,
+  });
+
+export const Instrument = ({
+  transactionName = '',
+  buildTransactionIdSuffix,
+}: InstrumentationOptions = {}): any =>
+  instrumentationWrapper({
+    transactionName,
+    instrumentationType: 'Function',
+    buildTransactionIdSuffix,
+  });
+
+function instrumentationWrapper({
+  transactionName = '',
+  instrumentationType = 'Function',
+  buildTransactionIdSuffix,
+}: InstrumentationOptions & { instrumentationType: string }): any {
+  return (target: any, key: any, descriptor: PropertyDescriptor): any => {
+    const method = descriptor.value;
+    const methodName = transactionName || key;
+
+    const transactionIdentifierBase = `${instrumentationType}/${target?.constructor?.name}/${methodName}`;
+
+    let nr: any = null;
+    try {
+      // Dynamically load newrelic
+      // eslint-disable-next-line global-require
+      nr = require('newrelic');
+    } catch {
+      return descriptor;
+    }
+
+    if (nr) {
+      const isAsync = method.constructor.name === 'AsyncFunction';
+
+      if (!isAsync) {
+        // eslint-disable-next-line no-param-reassign
+        descriptor.value = function (...args) {
+          const customSuffix = buildTransactionIdSuffix
+            ? `:${buildTransactionIdSuffix(...args)}`
```

Comment:
What do we think of the colon `:` delimiter? We could alternately use the same slash `/` delimiter as used on `transactionIdentifierBase`, but I chose a different delimiter to differentiate the suffixed part.